flaky test: `TestELSyncSafeRetractedByOffset` — CrossSafe sync timeout during preset setup

**Claude:** Reporting a new flake.

## Flaky Test

`TestELSyncSafeRetractedByOffset` in `op-acceptance-tests/tests/sync/elsync/offset_el_safe/`

## CI Link

https://circleci.com/gh/ethereum-optimism/optimism/4907858

- Job: `memory-all-opn-op-geth` (CircleCI 4907858)
- Workflow: `main`
- Merge-queue pipeline for PR #20250, commit `c409fb370e462945e77526dd63afdaf87c4413c8`
- Test duration: 72.06s (of 940.43s total job)

## Error

```
--- FAIL: TestELSyncSafeRetractedByOffset (72.06s)
    assertions.go:387:
        Error Trace:  /home/circleci/project/op-devstack/dsl/check.go:26
                      /home/circleci/project/op-devstack/presets/singlechain_from_runtime.go:51
                      /home/circleci/project/op-devstack/presets/singlechain_multinode.go:22
                      /home/circleci/project/op-acceptance-tests/tests/sync/elsync/offset_el_safe/init_test.go:15
                      /home/circleci/project/op-acceptance-tests/tests/sync/elsync/offset_el_safe/sync_test.go:28
        Error: Received unexpected error:
               operation failed permanently after 30 attempts: expected head to match: safe
```

## Context

The failure is during **preset setup**, before the test body runs — `singlechain_from_runtime.go:51` is the preset's initial CrossSafe/LocalUnsafe sync check, identical symptom to #19897. The preset waits up to 30×2s = 60s for the verifier CL to reach `types.CrossSafe` matching the sequencer.

The preset in `init_test.go` configures `VerifierSyncMode = nodeSync.ELSync` + `OffsetELSafe = 10s` for both L2CLs globally. In ELSync mode, `CrossSafe` requires derivation + supervisor confirmation of old batched data — slower than blocks arriving via EL p2p.

## Related

- #19897 — `TestTruncateDatabaseOnELResync`: same preset hook, same error message, also ELSync mode. Likely same root cause (CrossSafe promotion slower than 60s budget when under CI resource contention).

## Date

2026-04-24

--- a/op-devstack/presets/singlechain_from_runtime.go
+++ b/op-devstack/presets/singlechain_from_runtime.go
@@ -5,8 +5,27 @@ import (
 	"github.com/ethereum-optimism/optimism/op-devstack/dsl"
 	"github.com/ethereum-optimism/optimism/op-devstack/stack"
 	"github.com/ethereum-optimism/optimism/op-devstack/sysgo"
+	nodeSync "github.com/ethereum-optimism/optimism/op-node/rollup/sync"
 	"github.com/ethereum-optimism/optimism/op-service/eth"
 	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+)
+
+// Initial-sync-check budgets for singleChainMultiNodeFromRuntime. Attempts are
+// polled every 2s by dsl.MatchedFn.
+//
+//   - LocalUnsafe: EL P2P / CL gossip is fast; 60s covers CI contention.
+//   - CrossSafe (CLSync mode): requires derivation to process the first L1
+//     origin and advance safe; 60s is enough in practice.
+//   - CrossSafe (ELSync mode): the verifier must first complete EL sync over
+//     P2P (which can take tens of seconds on a cold node under CI load)
+//     before the forkchoice update that seeds safe/finalized fires; only
+//     after that does derivation start and begin advancing CrossSafe. The
+//     combined path is consistently slower than the CLSync case, so we give
+//     it a 4x budget. See issue #20334 for the failure mode this guards.
+const (
+	initialSyncCheckAttemptsLocalUnsafe     = 30
+	initialSyncCheckAttemptsCrossSafe       = 30
+	initialSyncCheckAttemptsCrossSafeELSync = 120
 )
 
 func singleChainMultiNodeFromRuntime(t devtest.T, runtime *sysgo.SingleChainRuntime, runSyncChecks bool) *SingleChainMultiNode {
@@ -48,9 +67,16 @@ func singleChainMultiNodeFromRuntime(t devtest.T, runtime *sysgo.SingleChainRunt
 	}
 	if runSyncChecks {
 		// Ensure the follower node is in sync with the sequencer before starting tests.
+		// CrossSafe requires derivation to run, which under ELSync can only begin
+		// after the EL completes P2P sync and the node emits its first forkchoice
+		// update — so size that budget by the follower's sync mode.
+		crossSafeAttempts := initialSyncCheckAttemptsCrossSafe
+		if opNode, ok := nodeB.CL.(*sysgo.OpNode); ok && opNode.SyncMode() == nodeSync.ELSync {
+			crossSafeAttempts = initialSyncCheckAttemptsCrossSafeELSync
+		}
 		dsl.CheckAll(t,
-			preset.L2CLB.MatchedFn(preset.L2CL, types.CrossSafe, 30),
-			preset.L2CLB.MatchedFn(preset.L2CL, types.LocalUnsafe, 30),
+			preset.L2CLB.MatchedFn(preset.L2CL, types.CrossSafe, crossSafeAttempts),
+			preset.L2CLB.MatchedFn(preset.L2CL, types.LocalUnsafe, initialSyncCheckAttemptsLocalUnsafe),
 		)
 	}
 	return preset

--- a/op-devstack/sysgo/l2_cl_opnode.go
+++ b/op-devstack/sysgo/l2_cl_opnode.go
@@ -7,6 +7,7 @@ import (
 	"github.com/ethereum-optimism/optimism/op-devstack/devtest"
 	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/opnode"
 	"github.com/ethereum-optimism/optimism/op-node/config"
+	nodeSync "github.com/ethereum-optimism/optimism/op-node/rollup/sync"
 	"github.com/ethereum-optimism/optimism/op-service/clock"
 	"github.com/ethereum-optimism/optimism/op-service/eth"
 	"github.com/ethereum-optimism/optimism/op-service/testutils/tcpproxy"
@@ -22,11 +23,18 @@ type OpNode struct {
 	interopEndpoint  string
 	interopJwtSecret eth.Bytes32
 	cfg              *config.Config
+	syncMode         nodeSync.Mode
 	p                devtest.CommonT
 	logger           log.Logger
 	userProxy        *tcpproxy.Proxy
 	interopProxy     *tcpproxy.Proxy
 	clock            clock.Clock
+}
+
+// SyncMode reports the sync mode the op-node was started with
+// (VerifierSyncMode for verifiers, SequencerSyncMode for sequencers).
+func (n *OpNode) SyncMode() nodeSync.Mode {
+	return n.syncMode
 }
 
 var _ L2CLNode = (*OpNode)(nil)

--- a/op-devstack/sysgo/singlechain_build.go
+++ b/op-devstack/sysgo/singlechain_build.go
@@ -430,12 +430,13 @@ func startL2CLNode(
 		ExperimentalOPStackAPI:          true,
 	}
 	l2CL := &OpNode{
-		name:   startCfg.Key,
-		opNode: nil,
-		cfg:    nodeCfg,
-		p:      t,
-		logger: logger,
-		clock:  clock.SystemClock,
+		name:     startCfg.Key,
+		opNode:   nil,
+		cfg:      nodeCfg,
+		syncMode: syncMode,
+		p:        t,
+		logger:   logger,
+		clock:    clock.SystemClock,
 	}
 	l2CL.Start()
 	t.Cleanup(l2CL.Stop)

